### PR TITLE
Range-ify std.string.center() with centerJustifier()

### PR DESCRIPTION
**Pull request by WalterBright**
http://dlang.org/phobos/std_string.html#.center

100% unit test coverage

### CODE REVIEW

**Comment by kuettler** [comment]
This is very fine code. By no means I am here to criticize. Instead, I ask to learn.

The code is overly verbose for my liking. Dogfooding on phobos, I'd rather write the inner section thusly:
```
else static if (C.sizeof == 4)
{
    auto len = maxRangeLength(r, width);
    auto nleft = (width - len) / 2;
    auto nright = width - len - nleft;
    return chain(repeat(fillChar, nleft), r, repeat(fillChar, nright));
}
```
where `maxRangeLength` does the lookahead, if required.

So why is the above inferior to the explicit code in this PR?

**Comment by WalterBright** [comment]
@kuettler very nice. I should have thought of it. My only criticism of it is the initialization is eager, rather than lazy. (I.e. maxRangeLength potentially runs through the range.) I'd like to keep the initialization trivial and lazy.

**Comment by WalterBright** [comment]
rebased

**Comment by WalterBright** [comment]
Implemented @kuettler 's suggestion for ranges with hasLength.

**Comment by kuettler** [comment]
@WalterBright Thank you. I did not expect that.

You are right about lazy initialization. I would use the two argument version of `walkLength` in `initialize` to get rid of the explicit loop. This might be a matter of taste.

Two minor questions for me to learn: (a) Some variables in the `Result` struct start with an underscore, others do not. What is the rationale? Do you consider some variables as part of the interface? (b) What is the reason for the alias from `_width` to `nleft`? From my perspective it complicates the code.

File: std/string.d [hunk 1]
```diff
@@ -3215,6 +3186,202 @@ S center(S)(S s, size_t width, dchar fillChar = ' ') @trusted pure
...
+                    if (!_input.empty)
+                        _input.popFront();  // fast path
+                    else
+                        --nright;
+                }
+                else
+                {
+                    if (!inited)
+                        initialize();
+                    if (nleft)
```
Comment (jwhear) [inline std/string.d]: Isn't this branch guaranteed given outer if..else?

Comment (kuettler) [inline std/string.d]: No, `initialize()` might change the state. This is exactly why I like using ranges much more than writing them.

Comment (jwhear) [inline std/string.d]: Doh.  Perhaps a comment is in order here.

**Comment by WalterBright** [comment]
> (a) Some variables in the  Result  struct start with an underscore, others do not. What is the rationale? Do you consider some variables as part of the interface?

There isn't really a rationale, other than they need a name that is same-only-different from the parameter.

> (b) What is the reason for the alias from  _width  to  nleft ? From my perspective it complicates the code.

I tried using a union, but CTFE fails on unions. The purpose of them occupying the same memory space is to minimize stack space consumed. There are different names because they are used for different purposes.

If they were variables, rather than fields, the optimizer would recognize that their usages were disjoint and would map them to the same stack slot, but alas, this does not happen for fields.

I would say this is a bit of an advanced technique, probably not a good idea for general code, but Phobos code should be tuned for max performance.

**Comment by kuettler** [comment]
Thanks a lot for your explanation. This is very helpful.

I would tend to avoid the alias and use one (misnamed) variable instead, since both names are a little too close in lines 3263 to 3265. Also, lines 3289 and 3304 use `nleft` that really is `_width` before initialization. Yet, this is a matter of preference again.

Thus, LGTM.

File: std/string.d [hunk 2]
```diff
@@ -3215,6 +3186,202 @@ S center(S)(S s, size_t width, dchar fillChar = ' ') @trusted pure
...
+        import std.utf : byDchar, byWchar;
+        return centerJustifier(r.byDchar, width, fillChar).byWchar;
+    }
+    else static if (C.sizeof == 4 && hasLength!Range)
+    {
+        import std.range : chain, repeat;
+
+        auto len = r.length;
+        if (len > width)
+            len = width;
```
Comment (MartinNowak) [inline std/string.d]: You can use `walkLength(r, up_to=width)` for ranges without length, then the explicit struct version below can go.

**Comment by WalterBright** [comment]
rebased

File: std/string.d (hunk 2 above)
Comment (WalterBright) [inline std/string.d]: You're right. Nice.

**Comment by andralex** [comment]
noice

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**